[12.0] Backport of 13.0 multiple improvements

etobella opened the pull request:
Backport of #305 

@simahawk @guewen

Commit:
Add job_serialized in IrModelFields.ttype selection

Context
-------

The queue_job module adds a new type of field, with the type
"job_serialized", which is close to the base "serialized" field,
adding support of serialization for recordsets and a couple of common
types such as datetime and date.

This field type is used in several fields of the "queue.job" model.

In "ir.model.fields", each field is related to a "ttype", which is
a selection containing every types of fields supported by the current
database.

The "job_serialized" field was never registered as a possible value for
the ttype selection field. It seems it had never been an issue before
Odoo 13.0. With Odoo 13.0, we may have this error on upgrades:

    INFO sample odoo.modules.loading: 297 modules loaded in 529.51s, 0 queries
    INFO sample odoo.addons.base.models.ir_model: Deleting 4@ir.model.fields.selection (base.selection__ir_model_fields__ttype__job_serialized)
    ERROR sample odoo.sql_db: bad query: UPDATE "ir_model_fields" SET "ttype"=NULL WHERE "ttype"='job_serialized'
    ERROR: null value in column "ttype" violates not-null constraint
    DETAIL:  Failing row contains (4296, record_ids, null, queue.job, null, null, null, 266, Record, null, null, t, null, null, f, t, f, f, null, base, null, null, t, null, null, null, null, null, t, null, null, 1, 2020-08-11 13:03:09.913127, null, null).

    2020-08-11 13:10:53,168 167 WARNING sample odoo.modules.loading: Transient module states were reset
    2020-08-11 13:10:53,170 167 ERROR sample odoo.modules.registry: Failed to load registry
    Traceback (most recent call last):
      File "/odoo/src/odoo/modules/registry.py", line 86, in new
        odoo.modules.load_modules(registry._db, force_demo, status, update_module)
      File "/odoo/src/odoo/modules/loading.py", line 472, in load_modules
        env['ir.model.data']._process_end(processed_modules)
      File "/odoo/src/odoo/addons/base/models/ir_model.py", line 1990, in _process_end
        record.unlink()
      File "/odoo/src/odoo/addons/base/models/ir_model.py", line 1208, in unlink
        self.env.cr.execute(query, [selection.value])
      File "/odoo/src/odoo/sql_db.py", line 168, in wrapper
        return f(self, *args, **kwargs)
      File "/odoo/src/odoo/sql_db.py", line 245, in execute
        res = self._obj.execute(query, params)
    psycopg2.IntegrityError: null value in column "ttype" violates not-null constraint
    DETAIL:  Failing row contains (4296, record_ids, null, queue.job, null, null, null, 266, Record, null, null, t, null, null, f, t, f, f, null, base, null, null, t, null, null, null, null, null, t, null, null, 1, 2020-08-11 13:03:09.913127, null, null).

Explanation
-----------

In 13.0, a model "ir.model.fields.selection" storing the allowed
selection values for a field has been added
(https://github.com/odoo/odoo/commit/7593b887dff4fcd474ca72aead6825481b419d86).

When the "job_serialized" fields for queue_job are created, they are
properly created in "ir.model.fields" with "job_serialized" as "ttype".
It creates the entry in "ir.model.fields.selection".

When we update the module, if it does not find "job_serialized" as a
possible value for IrModelFields.ttype, it unlinks the entry in
"ir.model.fields.selection". When this happens, an override of the
"unlink()" method in "ir.model.fields.selection" forces a NULL value in
the selection fields (here IrModelFields.ttype).

Fixes: #227

Commit:
Store worker's PID when a job is started

Add a field in the "queue_job" model to store the worker pid.
No usage for new besides "logging" for debugging (helps to find which
process to investigate in case a job is stuck).
Later, it may be useful to improve the recovery of job that stay in the
wrong state when the job runner restarts.

Closes #177

nilshamerlinck committed:
[ADD] check that queue_job_notify trigger exists

nilshamerlinck committed:
[ADD] queue_job: innocuous test job for debugging/monitoring purposes

PCatinean committed:
Added etree._Element to Json encoder/decoder

Commit:
Add method to patch a method to be automatically delayed

This patch method has to be called in ``_register_hook``.

When a method is patched, any call to the method will not directly
execute the method's body, but will instead enqueue a job.

When a ``context_key`` is set when calling ``_patch_job_auto_delay``,
the patched method is automatically delayed only when this key is
``True`` in the caller's context. It is advised to patch the method
with a ``context_key``, because making the automatic delay *in any
case* can produce nasty and unexpected side effects (e.g. another
module calls the method and expects it to be computed before doing
something else, expecting a result, ...).

A typical use case is when a method in a module we don't control is called
synchronously in the middle of another method, and we'd like all the calls
to this method become asynchronous.

It relies on https://github.com/OCA/queue/pull/274 that deprecates the
`@job` decorator.

Commit:
Refactor user_id,model_name,record_ids fields on queue.job

These fields contain the current model and record ids and user id the
job works on.
Storing them as separate fields does not allow to take benefits on the
"JobSerialized" field, which keeps the original user, and later the 'su'
flag. We could store the context there as well (or a cleaned one).

Use computed fields to keep the model_name, record_ids, user_id fields for
backward compatibility. The model_name should stay anyway for the
searches on the UI, and the user_id for the same reason + possibility to
edit the user.

Commit:
Deprecate job decorators

* @job and @related_action are still supported but emit warnings and
  will be removed at some point (code to remove has a TODO with
  :job-no-decorator:)
* the options previously set by the decorators are now stored on
  "queue.job.function" records, which can be explicitly created
  as XML records (and then editable!)
* any method can be used with "with_delay()", if no "queue.job.function"
  exists, default values are used (root channel, default related
  action, ...)

Commit:
Rebind job functions/channels with same name during install

As previously, the job functions and channels were created automatically
by "_register_job" on methods decorated by `@job`, databases have
records without xmlids.

To prevent duplicates (or unique constraint error) in case of channels,
the create method automatically merges the new record with the existing
record. This behavior happens only when installing/upgrading a module,
records created manually behave normally.

If 2 modules adds the same channel or job function, they'll be merged
together with 2 xmlids, on uninstallation of one of the module, the
record will be kept thanks to the second xmlid.

Commit:
Update documentation for deprecation of decorators

Commit:
Add safety to prevent crafting queue.job records

As we can delay a job on any method, and the queue.job model is
accessible from RPC (as any model), prevent to:

* create a queue.job using RPC
* write on protected fields (e.g. method name) using RPC

Admittedly, the risk is low since users need have Queue Job Manager
access to create/write on jobs, but it would allow these users to call
internal methods.

The check is done using a context key that must be equal to a sentinel
object, which is impossible to pass through RPC.

Commit:
Replace job function name by model + method fields

A many2one to select the model + a char for the method name is more
convenient and error-proof than having to type the expected format
correctly.

Commit:
Add missing sudo() when reading queue.job.function

Users shouldn't need access to queue.job.function.
Read config as sudo().

Commit:
Remove warning when @job decorator is used

It makes the runbot builds fail. There is no urgency to remove the
decorator even if we should incentivize the removal because it is
no longer supported, hence the info log. Also, show only the XML snippet
when debug logs are active, because it is too verbose.

Commit:
Store 'su' flag in records stored in JobSerialized fields

Which means that the record the job is working on, or any record
argument of the job keeps the 'su' flag.

atchuthan committed:
[FIX] queue_job: USAGE update

Commit:
Fix required sudo in Job

Following changes of https://github.com/OCA/queue/pull/281
The initial sudo() is lost when we call "with_env()" with a False su
flag. Ensure the read job.record keeps a su flag.

nilshamerlinck committed:
[FIX] run doctests as part of standard tests

Commit:
Remove initial create notification and follower

Everytime a job is created, a mail.message "Queue Job created" is
created. This is useless, as we already have the creation date and user,
and nobody will ever want to receive a notification for a created job
anyway.

Removing the on creation auto-subscription of the user that created the
job makes sense as well since we automatically subscribe the queue job
managers for failures. Add the owner of the jobs to the followers on
failures only as well.

It allows to remove a lot of insertions of records (and of deletions
when autovacuuming jobs).

Commit:
Change technical fields to read-only

These fields should not be changed by users.

Commit:
Optimize queue.job creation

Several fields on queue.job are initialized using computed fields,
then never changed again. On creation of a queue.job record, we'll have
an initial INSERT + at least one following UPDATE for the computed
fields.

Replace all the stored computed fields by a raw initialization of the
values in `Job.store()` when the job is created, so we have only a
single INSERT.

Highlights:

* as channel is no longer a compute/inverse field, override_channel is
  useless, I dropped it (actually the same value was stored in both
  channel and override_channel as the channel field was stored)
* one functional diff is that now, when changing a channel on a
  job.function, the channel is no longer synchronized on existing jobs,
  it will be applied only on new jobs: actually this is an improvement,
  because it was impossible to change the channel of a job function
  in a large queue_job table as it meant writing on all the done/started
  jobs
* searching the queue.job.function is now cached, as each job using the
  same will run a query on queue_job_function

simahawk committed:
queue_job: add exec time to view some stats

Commit:
Add model in search view / group by

Commit:
Fix date_done set when state changes back to pending

When Job.date_done has been set, for instance because the job has been
performed, if the job is set back to pending (e.g. a RetryableJobError
is raised), the date_done is kept.

simahawk committed:
queue_job: close buffer when done

simahawk committed:
queue_job: store exception name and message

simahawk committed:
queue_job: improve filtering and grouping

simahawk committed:
queue_job: add hook to customize stored values

simahawk committed:
queue_job: migration step to store exception data

etobella commented:
Also added #312 #310 and #309

simahawk reviewed (approved):
just code review, LG

guewen reviewed (approved):
Thanks a lot!

guewen commented:
Hi @sbidoul, @gurneyalex,

Runbot fails because of this on the PRs for 12.0:

```
Unable to find image 'oca-queue:fd5b8ac_env_1_job_2' locally
docker: Error response from daemon: pull access denied for oca-queue, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
```

Do you happen to know why? 😌

guewen commented:
Oh maybe I'm looking at the wrong logs and the problem would be phantom-js.

sbidoul commented:
Could it be the new(ish) docker hub rate limits that kick in ?

StefanRijnhart commented:
I believe you can squash out the commit from #312 because that was a port forward of #311

StefanRijnhart commented on queue_job/job.py:
```diff
@@ -1,32 +1,32 @@
-from .exception import (NoSuchJobError,
-                        FailedJobError,
-                        RetryableJobError)
+from .exception import FailedJobError, NoSuchJobError, RetryableJobError
```
What do you think about undoing the changes from python-black in this PR? It would make it a lot easier to review.

StefanRijnhart commented on queue_job/controllers/main.py:
```diff
@@ -127,3 +113,35 @@ def retry_postpone(job, message, seconds=None):
 
         return ""
+
+    @http.route("/queue_job/create_test_job", type="http", auth="user")
```
This is strange. In 13.0, this code is added by https://github.com/OCA/queue/pull/265 by @nilshamerlinck (and as such also present in https://github.com/OCA/queue/pull/330), but in your git history it is attributed to @guewen as part of https://github.com/OCA/queue/pull/333/commits/7385fd54fcbb2370413fb1fb305b6f213f8150a2.

I'm getting the feeling that maybe these backports should be split up in separate PRs.

etobella replied:
I think I did something wrong with the commits. I will recheck thanks :wink:

etobella committed:
[BPRT] Final issues for the backport

etobella commented:
I think all the missing commits are here
Also, all the changes from black shouldn't be here.

StefanRijnhart commented:
It's gotten bigger again, and do we really want to backport those large API changes to this older release?

etobella commented:
It is bigger because I analyzed all the commits that where not backported. Originally, it was intended to backport them (maybe not all of them, but a lot), but it wasn't done.

etobella commented:
@OCA/connector-maintainers 
The issue in runbot already exists on 12.0 branch of queue
https://runbot.odoo-community.org/runbot/repo/https-github-com-oca-queue-230

guewen commented:
I tried to use this branch on a project and the migration failed, going to investigate why.

guewen commented on queue_job/migrations/12.0.2.3.0/post-migration.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def _get_exception_details(job):
+    for line in reversed(job.exc_info.splitlines()):
```
I had to do this: 

```suggestion
    exc_info = job.read(fields=['exc_info'])[0]['exc_info']
    for line in reversed(exc_info.splitlines()):
```

Otherwise, browsing the job could fail if the job is on a model which is not yet loaded by the registry:

```
Traceback (most recent call last):
  File "/odoo/src/odoo/service/server.py", line 1162, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/odoo/src/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/odoo/src/odoo/modules/loading.py", line 417, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/odoo/src/odoo/modules/loading.py", line 313, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/odoo/src/odoo/modules/loading.py", line 227, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "/odoo/src/odoo/modules/migration.py", line 190, in migrate_module
    migrate(self.cr, installed_version)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 14, in migrate
    _compute_jobs_new_values(env)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 21, in _compute_jobs_new_values
    exception_details = _get_exception_details(job)
  File "/odoo/external-src/queue/queue_job/migrations/12.0.2.3.0/post-migration.py", line 27, in _get_exception_details
    for line in reversed(job.exc_info.splitlines()):
  File "/odoo/src/odoo/fields.py", line 1036, in __get__
    self.determine_value(record)
  File "/odoo/src/odoo/fields.py", line 1139, in determine_value
    record._prefetch_field(self)
  File "/odoo/src/odoo/models.py", line 2880, in _prefetch_field
    result = records.read([f.name for f in fs], load='_classic_write')
  File "/odoo/src/odoo/models.py", line 2835, in read
    vals[name] = convert(record[name], record, use_name_get)
  File "/odoo/src/odoo/models.py", line 5167, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/odoo/src/odoo/fields.py", line 1044, in __get__
    return self.convert_to_record(value, record)
  File "/odoo/external-src/queue/queue_job/fields.py", line 64, in convert_to_record
    return json.loads(value or default, cls=JobDecoder, env=record.env)
  File "/usr/lib/python3.5/json/__init__.py", line 332, in loads
    return cls(**kw).decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/odoo/external-src/queue/queue_job/fields.py", line 107, in object_hook
    model = self.env[obj['model']]
  File "/odoo/src/odoo/api.py", line 831, in __getitem__
    return self.registry[model_name]._browse((), self)
  File "/odoo/src/odoo/modules/registry.py", line 177, in __getitem__
    return self.models[model_name]
KeyError: 'xxx.product.template'
```

guewen replied:
But we would better have to read all the "exc_info" at a time for all the jobs rather than 1 by 1 as I did here.

guewen replied:
Here is the better proposal reading all the rows at once: https://github.com/OCA/queue/pull/333/files#r626439498

guewen commented:
Here's the correction for the issue I mentioned before https://github.com/OCA/queue/pull/333/files#r626409959

guewen commented on queue_job/migrations/12.0.2.3.0/post-migration.py:
```diff
@@ -0,0 +1,47 @@
+        _compute_jobs_new_values(env)
+
+
+def _compute_jobs_new_values(env):
+    for job in env["queue.job"].search(
+        [("state", "=", "failed"), ("exc_info", "!=", False)]
+    ):
+        exception_details = _get_exception_details(job)
+        if exception_details:
+            job.update(exception_details)
+
+
+def _get_exception_details(job):
+    for line in reversed(job.exc_info.splitlines()):
+        if _find_exception(line):
+            name, msg = line.split(":", 1)
+            return {
+                "exc_name": name.strip(),
+                "exc_message": msg.strip("()', \""),
+            }
```
Better one reading all the exc_info at once:
```suggestion
def _compute_jobs_new_values(env):
    job_exc_infos = env["queue.job"].search(
        [("state", "=", "failed"), ("exc_info", "!=", False)]
    ).read(fields=["exc_info"])
    for job_row in job_exc_infos:
        exception_details = _get_exception_details(job_row["exc_info"])
        if exception_details:
            job = env["queue.job"].browse(job_row["id"])
            job.write(exception_details)


def _get_exception_details(exc_info):
    for line in reversed(exc_info.splitlines()):
        if _find_exception(line):
            name, msg = line.split(":", 1)
            return {
                "exc_name": name.strip(),
                "exc_message": msg.strip("()', \""),
            }
```

guewen replied:
Hi @etobella, could you merge this suggestion?

etobella replied:
Sure!!! Done!

guewen replied:
Thanks :)

etobella committed:
[FIX] queue_job: Fix migration script

guewen commented:
Hi @StefanRijnhart, do you still have concerns or can we merge the backports?

StefanRijnhart commented:
For me this is too much to review at once, and to assess the impact of. But I do trust you fully with that, so don't mind me ;-)

guewen committed:
Fix display on exec_time on tree view as seconds

The float_time widget shows hours seconds, we only want seconds.
The widget had been removed on the form view, but not on the tree view.

guewen commented on queue_job/views/queue_job_views.xml:
```diff
@@ -87,15 +99,39 @@
                 <field name="date_created"/>
+                <field name="eta"/>
                 <field name="date_done"/>
+                <field name="exec_time" widget="float_time" />
```
```suggestion
                <field name="exec_time" />
```

@etobella can you apply this change please? (I did the same for 13.0 in #352)

guewen commented:
@etobella do you run this on production already?
FYI, as the company I work for is on 12.0 and is a large consumer of jobs, I should deploy this branch soon to put this to test before merging.

etobella commented:
I did the backport in order to backport `edi`, but at the end we did not deploy the change. Sorry :cry:

guewen commented:
> I did the backport in order to backport `edi`, but at the end we did not deploy the change. Sorry

No problem, I only asked to assess the risks ;)

nilshamerlinck commented:
Hi, fyi I've been running this branch in production for 2 weeks without issue.

guewen commented:
> Hi, fyi I've been running this branch in production for 2 weeks without issue.

Thanks for your feedback @nilshamerlinck!
Running well on the test (using production database) service here too.

Thanks @etobella for the backport :)

/ocabot merge major

pedrobaeza commented:
Since the merge of this, I get continuous logs with this problem, and no job is executed:
<details>

```
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/serving.py", line 270, in run_wsgi
odoo_1    |     execute(self.server.app)               
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/serving.py", line 258, in execute
odoo_1    |     application_iter = app(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 142, in application
odoo_1    |     return application_unproxied(environ, start_response)                                                                                                                                
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/wsgi_server.py", line 117, in application_unproxied
odoo_1    |     result = odoo.http.root(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1320, in __call__
odoo_1    |     return self.dispatch(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1293, in __call__
odoo_1    |     return self.app(environ, start_wrapped)
odoo_1    |   File "/home/odoo/.local/lib/python3.5/site-packages/werkzeug/wsgi.py", line 766, in __call__
odoo_1    |     return self.app(environ, start_response)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 1488, in dispatch
odoo_1    |     result = ir_http._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/website_sale/models/ir_http.py", line 15, in _dispatch
odoo_1    |     return super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/auth_signup/models/ir_http.py", line 19, in _dispatch
odoo_1    |     return super(Http, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/web_editor/models/ir_http.py", line 22, in _dispatch
odoo_1    |     return super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/http_routing/models/ir_http.py", line 403, in _dispatch
odoo_1    |     result = super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/utm/models/ir_http.py", line 26, in _dispatch
odoo_1    |     response = super(IrHttp, cls)._dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 212, in _dispatch
odoo_1    |     return cls._handle_exception(e)
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/website/models/ir_http.py", line 195, in _handle_exception
odoo_1    |     return super(Http, cls)._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/addons/utm/models/ir_http.py", line 31, in _handle_exception
odoo_1    |     response = super(IrHttp, cls)._handle_exception(exc)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 182, in _handle_exception
odoo_1    |     return request._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 776, in _handle_exception
odoo_1    |     return super(HttpRequest, self)._handle_exception(exception)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 314, in _handle_exception
odoo_1    |     raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 87, in reraise
odoo_1    |     raise value
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 208, in _dispatch
odoo_1    |     result = request.dispatch()
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 835, in dispatch
odoo_1    |     r = self._call_function(**self.params)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 346, in _call_function
odoo_1    |     return checked_call(self.db, *args, **kwargs)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 98, in wrapper
odoo_1    |     return f(dbname, *args, **kwargs)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
odoo_1    |     result = self.endpoint(*a, **kw)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 941, in __call__
odoo_1    |     return self.method(*args, **kw)
odoo_1    |   File "/opt/odoo/custom/src/odoo/odoo/http.py", line 519, in response_wrap
odoo_1    |     response = f(*args, **kw)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/controllers/main.py", line 81, in runjob
odoo_1    |     job = Job.load(env, job_uuid)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/job.py", line 252, in load
odoo_1    |     return cls._load_from_db_record(stored)
odoo_1    |   File "/opt/odoo/custom/src/queue/queue_job/job.py", line 263, in _load_from_db_record
odoo_1    |     method = getattr(recordset, method_name)
odoo_1    | AttributeError: 'base' object has no attribute 'confirm_one_invoice' - - -
```

</details>

Can you point which can be the problem?

guewen commented:
@pedrobaeza can you paste the func_string, method_name, and records, args, kwargs json fields of one of these jobs?

pedrobaeza commented:
Thanks for the quick reply. It's this function:

https://github.com/OCA/l10n-spain/blob/5c02c1d9caf61e02f57e2f2a4ec4c675283e92de/l10n_es_aeat_sii/models/account_invoice.py#L1450

and the job is this way:

![Selección_030](https://user-images.githubusercontent.com/7165771/124515015-a96e7a00-ddde-11eb-8941-de2528474efc.png)

This is before switching to new version.

Seeing 13.0 version, it seems to not be any difference:

https://github.com/OCA/l10n-spain/blob/23924f08218034c5f6b8f92ed2034e500070dd42/l10n_es_aeat_sii_oca/models/account_move.py#L1452

StefanRijnhart commented:
Was this job created before you updated the code of the module and ran the module upgrade?

It looks like https://github.com/OCA/queue/pull/333/files#diff-1c5ee7d63e3224474074dbbdaa46e8d7a838199d2b4d4182726de22fd6e627d1R15-R16 might have been better as an end-migration script, because it might depend on the models and the methods being loaded at the time of migration.

If you query the database for the columns @guewen mentions, we can check this hypothesis.

-- edit -- that method looks safe though https://github.com/OCA/queue/blob/12.0/queue_job/models/queue_job.py#L563-L573

pedrobaeza commented:
Yeah, it can be. For now, I have go back with the previous commit. As soon as I have a slot, I will propose a PR changing it to end-migration and try again to see if that's the problem.

StefanRijnhart commented:
@pedrobaeza not sure that that's it. I have an environment with a lot of jobs on which I can check the migration, so I will post back about that as well.
BTW you might have to do some reverse migration if you want to go back, at least on jobs created after running with the code from this PR.
And it would still be informative to have the SQL values of the func_string, method_name, and records, args, kwargs columns of one of the failing jobs.

guewen commented:
According to the screenshot above, the `func_string` is correct, so I don't think this migration is the problem.
According to the stack trace, the most likely is that this migration https://github.com/OCA/queue/blob/ad28e410e1d1c033f0df7810d086a134e210091d/queue_job/migrations/12.0.2.1.0/pre-migration.py#L20-L26 didn't run or for some reason generated a wrong json in the `records` field. The default value being the `base` model: https://github.com/OCA/queue/blob/ad28e410e1d1c033f0df7810d086a134e210091d/queue_job/fields.py#L33-L35

So:

1. Check if the migration 12.0.2.1.0 runs
2. Check and give feedback about the SQL content of the `records` field (func_string, method_name, args, kwargs may help but I really expect the issue to be in `records`).
